src/backend/distributed/executor/intermediate_result_encoder.c

```diff
@@ -0,0 +1,324 @@
+ */
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "postgres.h"
+#include "funcapi.h"
+#include "libpq-fe.h"
+#include "miscadmin.h"
+#include "pgstat.h"
+
+#include "catalog/pg_enum.h"
+#include "commands/copy.h"
+#include "distributed/commands/multi_copy.h"
+#include "distributed/connection_management.h"
+#include "distributed/intermediate_results.h"
+#include "distributed/master_metadata_utility.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/multi_client_executor.h"
+#include "distributed/multi_executor.h"
+#include "distributed/remote_commands.h"
+#include "distributed/transmit.h"
+#include "distributed/transaction_identifier.h"
+#include "distributed/tuplestore.h"
+#include "distributed/version_compat.h"
+#include "distributed/worker_protocol.h"
+#include "mb/pg_wchar.h"
+#include "nodes/makefuncs.h"
+#include "nodes/parsenodes.h"
+#include "nodes/primnodes.h"
+#include "storage/fd.h"
+#include "tcop/tcopprot.h"
+#include "utils/builtins.h"
+#include "utils/lsyscache.h"
+#include "utils/memutils.h"
+#include "utils/syscache.h"
+
+
+#define ENCODER_BUFFER_SIZE (4 * 1024 * 1024)
+
+/* internal state of intermediate result file encoder */
+struct IntermediateResultEncoder
+{
+	/*
+	 * Result of encoding is accumulated in outputBuffer, until it is flushed.
+	 * The flush is done when the data is returned as result of
+	 * IntermediateResultEncoderReceive() and IntermediateResultEncoderDone()
+	 * functions.
+	 */
+	StringInfo outputBuffer;
+
+	/*
+	 * Used for returning the flushed result of encoding, at which time move the
+	 * data pointer from outputBuffer to flushBuffer before reseting length of
+	 * outputBuffer.
+	 *
+	 * This is kept here to avoid allocating it everytime we need to flush some data.
+	 */
+	StringInfo flushBuffer;
+
+	IntermediateResultFormat format;
+	TupleDesc tupleDescriptor;
+
+	/* used when format is *_COPY_FORMAT */
+	CopyOutState copyOutState;
+	FmgrInfo *columnOutputFunctions;
+};
+
+/* forward declaration of local functions */
+static void ReadCopyFileIntoTupleStore(char *fileName, char *copyFormat,
+									   TupleDesc tupleDescriptor,
+									   Tuplestorestate *tupstore);
+static Relation StubRelation(TupleDesc tupleDescriptor);
+
+
+/*
+ * IntermediateResultEncoderCreate returns an encoder which can encode tuples of
+ * given tupleDesc form using the given format. Encoder possibly can add header
+ * data at this stage.
+ */
+IntermediateResultEncoder *
+IntermediateResultEncoderCreate(TupleDesc tupleDesc,
+								IntermediateResultFormat format,
+								MemoryContext tupleContext)
+{
+	IntermediateResultEncoder *encoder = palloc0(sizeof(IntermediateResultEncoder));
+	encoder->format = format;
+	encoder->tupleDescriptor = CreateTupleDescCopy(tupleDesc);
+	encoder->outputBuffer = makeStringInfo();
+	encoder->flushBuffer = makeStringInfo();
+
+	if (format == TEXT_COPY_FORMAT || format == BINARY_COPY_FORMAT)
```

Comment:
In this PR, this is always true. But when a new format is added, it is not always true anymore.